amplipi/auth.py

```diff
@@ -0,0 +1,216 @@
+import os
+import json
+import time
+import secrets
+
+from typing import Union, Dict
+from typing_extensions import Literal
+from datetime import datetime, timedelta, timezone
+
+from fastapi import Depends, FastAPI, HTTPException, status, APIRouter, Request
+from fastapi.responses import Response, RedirectResponse, FileResponse
+from fastapi.security import APIKeyCookie, APIKeyQuery, OAuth2PasswordRequestForm
+from fastapi.openapi.models import APIKey
+from fastapi.templating import Jinja2Templates
+from starlette.templating import _TemplateResponse as TemplateResponse
+from argon2 import PasswordHasher, Parameters as Argon2Params, Type as Argon2Type
+from hmac import compare_digest
+
+# pylint: disable=no-name-in-module
+from pydantic import BaseModel
+
+# The below takes ~200ms to hash on a RPi3 CM and is currently (fall 2023) recommended
+# as a 'safer' default by OWASP. These will only be used when instantiating a new hash;
+# existing hashes encode these parameters within the hash.
+argon2_params = Argon2Params(
+  time_cost=2,
+  memory_cost=19456,
+  parallelism=1,
+  type=Argon2Type.ID,
+  hash_len=32,
+  salt_len=16,
+  version=19
+)
+pwd_context = PasswordHasher.from_parameters(argon2_params)
+
+prefix = '/auth'
+
+router = APIRouter(prefix=prefix)
+
+templates = Jinja2Templates(directory="templates")
+
+USER_CONFIG_DIR = os.path.join(os.path.expanduser('~'), '.config', 'amplipi')
+
+USERS_FILE = os.path.join(USER_CONFIG_DIR, 'users.json')
+
+class UserData(BaseModel):
+  type: Literal["user", "api"]
+  access_key: Union[str, None]
+  access_key_updated: Union[datetime, None]
+  password_hash: Union[str, None]
+
+def get_users() -> dict:
+  """ Returns the users file contents """
+  users : Dict[str, Dict] = {}
+  # Load fields from users file (if it exists), falling back to no users.
+  try:
+    with open(USERS_FILE, encoding='utf-8') as users_file:
+      potential_users = json.load(users_file)
+      users.update(potential_users)
+  except FileNotFoundError as e:
+    print(f'Error loading users file: {e}')
+    print('Creating a users file from defaults.')
+    os.makedirs(USER_CONFIG_DIR, mode=0o700, exist_ok=True)
+    with open(USERS_FILE, encoding='utf-8', mode='w') as repair_file:
+      json.dump(users, repair_file)
+  except json.JSONDecodeError as e:
+    print(f'Error loading users file as JSON: {e}')
+    print('Moving the old one to a backup and creating a users file from defaults.')
+    os.rename(USERS_FILE, f"{USERS_FILE}.{time.time()}")
+    with open(USERS_FILE, encoding='utf-8', mode='w') as repair_file:
+      json.dump(users, repair_file)
+  except Exception as e:
+    print(f'Error loading identity file: {e}')
+    raise e
+  return users
```

Comment:
this entire function is pretty much copypasta from the `set_identity` function I created (but stopped using) in `utils.py`. There's a strong case to be made to move this into its own function, except there's also an aversion to tying this file into `utils.py` or any other location in this codebase, to permit the updater to function if a lot of the rest of the codebase is broken. thoughts?

Reply:
I think this makes sense as is for the most part. The only thing I would do is make set/get_users private so that we can change the format in the future without effecting code outside this file.